### PR TITLE
[DRM] Revert to single session decryption (when possible)

### PR DESCRIPTION
**Pull request by glennguy**
## Description

Many problems are now being observed from changes relating back to the introduction of CBCS decryption. For this change separate sessions were maintained, partly in an effort to avoid having more Bento4 patches. Now we observe the multiple license requests involved appear to inform service providers that there are multiple concurrent streams playing, causing license rejection after some duration of playback.

Another prior change was to force separate audio/video sessions on Android to resolve 'pixelation' issues. This workaround has worked but again as above has side effects. This issue should be investigated separately and a proper solution found.

## Motivation and context

In an effort to resolve issues experienced in #1147

## How has this been tested?

Runtime tests on CBCS and CENC test streams show no side effects. Artifacts of this PR will be distributed for user testing in particular on Android.

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the Wiki documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

**Comment by matthuisman** [comment]
looking at red lines vs green lines - seems this change simplifies the code quite a bit :)

Looking at your bento branch, seems the patch would be quite small?
Just this commit if im not mistaken: https://github.com/glennguy/Bento4/commit/3694e157a379199d29eebf5991723ba66db87b18

**Comment by matthuisman** [comment]
have tested with kayo

device 1) started windows stream - OK
device 2) started android 32bit stream - OK
device 3) started android 64bit stream - OK

once device .3 started - device .1 started getting the DRM errors (but kept playing)
Stopped device .1 and restarted it
Now device .2 starts getting DRM errors (but kept playing)

So seems Kayo just "kicks off" the oldest device when a new stream starts
Suprised device .2 still playing with drm error
Maybe it will stop after more time - will keep it playing

In summary  -works really well!! thank you so much @glennguy :)

**Comment by glennguy** [comment]
Thanks for testing @matthuisman !

Yes just changing the function signature is all that's needed in Bento4.

**Comment by CastagnaIT** [comment]
@matthuisman remember to never use `#` char to refer to a point on a list, because currently you have linked the PR to other issues/PR's ... we prefer to avoid it

**Comment by matthuisman** [comment]
@CastagnaIT Oops sorry

@glennguy will this be backported to Nexus?

**Comment by scottydulton** [comment]
> 

@matthuisman which 32 bit device did you test on?

**Comment by matthuisman** [comment]
Nvidia shield tube

**Comment by scottydulton** [comment]
@matthuisman still cant wrap my head around what tells the stream to parseandcorrect(from windows log) the url after license fail on all platforms except fire tv haha. and my fire TV log is very bare compared to windows log, far less information, is that normal? ive got all of the same component specific items enabled

**Comment by CastagnaIT** [comment]
i tried test this with n€tflix VP9.2 but something broken decrypting
https://paste.kodi.tv/atabitepoz.kodi

**Comment by glennguy** [comment]
I never checked the preinit code, will have a look today. I notice on the second licence call one of the keys comes back with a bad status code and that's the one that's failing

**Comment by glennguy** [comment]
@CastagnaIT would it be possible to get a copy of log where it works?

**Comment by CastagnaIT** [comment]
here https://paste.kodi.tv/mokivovitu.kodi

preinit dont have to make license call

File: src/Session.cpp [hunk 1]
```diff
@@ -450,26 +450,6 @@ bool CSession::InitializeDRM(bool addDefaultKID /* = false */)
       AP4_DataBuffer init_data;
```
Comment (CastagnaIT) [inline src/Session.cpp:466]: since the variable `currentSessionTypes` is not used anymore must be removed, on line 445 (gh dont allow to comment to unchanged code so i comment here)

**Comment by CastagnaIT** [comment]
the bento patch to `SetFragmentInfo` has also reversed method argument names, crypt vs skip blocks but seem to no affect the problem i found
however the bento patch added is irrilevant, the `SetFragmentInfo` is a custom method added by us that can be safetely removed
you can add the new virtual method directly to the `Adaptive_CencSingleSampleDecrypter`

**Comment by glennguy** [comment]
@CastagnaIT I think this is ready to go if you want to have another look? Working with NF now.

It won't be merged until done with your PR - reviewing that now as I type this. I will redo this to suit afterwards.

**Comment by CastagnaIT** [comment]
thanks
yes i will do a new review after that will be adapted with my rework PR

**Comment by matthuisman** [comment]
I suspect the big rework won't be backported? If this PR is updated for the rework, will this still be backported?

**Comment by glennguy** [comment]
> I suspect the big rework won't be backported?

Correct, Kodi Nexus is at release so we're not doing things of that nature in that branch, only bugfixes and feature backports if suitable

> If this PR is updated for the rework, will this still be backported?

Just done :) - #1216

**Comment by matthuisman** [comment]
oh great :) so i guess you backported the current changes and then update these changes to be compatible with rework PR :)
NICE!

File: src/Session.cpp [hunk 2]
```diff
@@ -459,34 +459,14 @@ bool CSession::InitializeDRM(bool addDefaultKID /* = false */)
...
-            sessionType = CPeriod::PSSHSet::MEDIA_VIDEO;
-            break;
-          case StreamType::AUDIO:
-            sessionType = CPeriod::PSSHSet::MEDIA_AUDIO;
-            break;
-          default:
-            break;
-        }
-      }
+      CPeriod::PSSHSet & sessionPsshset = m_adaptiveTree->m_currentPeriod->GetPSSHSets()[ses];
```
Comment (CastagnaIT) [inline src/Session.cpp]: clang format it

File: src/Session.cpp [hunk 3]
```diff
@@ -625,27 +605,34 @@ bool CSession::InitializeDRM(bool addDefaultKID /* = false */)
...
+            break;
           }
         }
+
       }
-      else if (defaultKid.empty() && !session.m_cencSingleSampleDecrypter)
+      else if (defaultKid.empty())
       {
-        LOG::Log(LOGWARNING, "Initializing stream with unknown KID!");
+        for (size_t i{ 1 }; i < ses; ++i)
```
Comment (CastagnaIT) [inline src/Session.cpp]: clang format it

File: src/common/AdaptiveCencSampleDecrypter.h [hunk 4]
```diff
@@ -0,0 +1,68 @@
+/*
+ *  Copyright (C) 2022 Team Kodi
```
Comment (CastagnaIT) [inline src/common/AdaptiveCencSampleDecrypter.h]: 2023

File: src/common/AdaptiveCencSampleDecrypter.h [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+class Adaptive_CencSampleDecrypter : public AP4_CencSampleDecrypter
+{
+public:
+  Adaptive_CencSampleDecrypter(AP4_CencSingleSampleDecrypter* single_sample_decrypter,
+                               AP4_CencSampleInfoTable* sample_info_table)
+    : AP4_CencSampleDecrypter(single_sample_decrypter, sample_info_table)
+  {
+  }
+
+  virtual AP4_Result DecryptSampleData(AP4_UI32 poolid,
```
Comment (CastagnaIT) [inline src/common/AdaptiveCencSampleDecrypter.h:20]: please move implementation in to relative .cpp file
also `#include "../utils/log.h"` must be moved to cpp

File: src/common/AdaptiveCencSampleDecrypter.h [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+                                       AP4_DataBuffer& data_in,
+                                       AP4_DataBuffer& data_out,
+                                       const AP4_UI08* iv)
+  {
+    // increment the sample cursor
+    unsigned int sample_cursor = m_SampleCursor++;
+
+    // setup the IV
+    unsigned char iv_block[16];
+    if (iv == NULL)
```
Comment (CastagnaIT) [inline src/common/AdaptiveCencSampleDecrypter.h]:
```suggestion
    if (!iv)
```

File: src/common/AdaptiveCencSampleDecrypter.h [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    {
+      iv = m_SampleInfoTable->GetIv(sample_cursor);
+    }
+    if (iv == NULL)
```
Comment (CastagnaIT) [inline src/common/AdaptiveCencSampleDecrypter.h]:
```suggestion
    if (!iv)
```

File: src/common/AdaptiveCencSampleDecrypter.h [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+      return AP4_ERROR_INVALID_FORMAT;
+    unsigned int iv_size = m_SampleInfoTable->GetIvSize();
+    AP4_CopyMemory(iv_block, iv, iv_size);
+    if (iv_size != 16)
+      AP4_SetMemory(&iv_block[iv_size], 0, 16 - iv_size);
+
+    // get the subsample info for this sample if needed
+    unsigned int subsample_count = 0;
+    const AP4_UI16* bytes_of_cleartext_data = NULL;
```
Comment (CastagnaIT) [inline src/common/AdaptiveCencSampleDecrypter.h]: use `nullptr`, also line above

File: src/common/AdaptiveDecrypter.h [hunk 9]
```diff
@@ -43,7 +36,31 @@ class Adaptive_CencSingleSampleDecrypter : public AP4_CencSingleSampleDecrypter
...
-  uint32_t m_CryptBlocks{0};
-  uint32_t m_SkipBlocks{0};
+  virtual AP4_Result SetFragmentInfo(AP4_UI32 pool_id,
+                                     const AP4_UI08* key,
+                                     const AP4_UI08 nal_length_size,
+                                     AP4_DataBuffer& annexb_sps_pps,
+                                     AP4_UI32 flags,
+                                     AP4_UI08 cryptoType,
+                                     AP4_UI08 cryptBlocks,
+                                     AP4_UI08 skipBlocks)
```
Comment (CastagnaIT) [inline src/common/AdaptiveDecrypter.h]: since we define an our method we can use `CryptoMode` to remove all casts?

```suggestion
  virtual AP4_Result SetFragmentInfo(AP4_UI32 pool_id,
                                     const AP4_UI08* key,
                                     const AP4_UI08 nal_length_size,
                                     AP4_DataBuffer& annexb_sps_pps,
                                     AP4_UI32 flags,
                                     CryptoMode cryptoMode,
                                     AP4_UI08 cryptBlocks,
                                     AP4_UI08 skipBlocks)
```

File: src/common/AdaptiveDecrypter.h [hunk 10]
```diff
@@ -43,7 +36,31 @@ class Adaptive_CencSingleSampleDecrypter : public AP4_CencSingleSampleDecrypter
...
+                                       const AP4_UI08* iv,
+                                       unsigned int subsample_count,
+                                       const AP4_UI16* bytes_of_cleartext_data,
+                                       const AP4_UI32* bytes_of_encrypted_data)
+  {
+    return AP4_CencSingleSampleDecrypter::DecryptSampleData(
+        data_in, data_out, iv, subsample_count, bytes_of_cleartext_data, bytes_of_encrypted_data);
+  };
+
+  virtual AP4_UI32 AddPool() { return 0; };
```
Comment (CastagnaIT) [inline src/common/AdaptiveDecrypter.h]:
```suggestion
  virtual AP4_UI32 AddPool() { return 0; }
```

File: src/common/AdaptiveDecrypter.h [hunk 11]
```diff
@@ -43,7 +36,31 @@ class Adaptive_CencSingleSampleDecrypter : public AP4_CencSingleSampleDecrypter
...
+  virtual void RemovePool(AP4_UI32 poolid){};
```
Comment (CastagnaIT) [inline src/common/AdaptiveDecrypter.h]:
```suggestion
  virtual void RemovePool(AP4_UI32 poolid) {}
```

File: src/common/AdaptiveDecrypter.h [hunk 12]
```diff
@@ -43,7 +36,31 @@ class Adaptive_CencSingleSampleDecrypter : public AP4_CencSingleSampleDecrypter
...
+  virtual const char* GetSessionId() { return nullptr; };
```
Comment (CastagnaIT) [inline src/common/AdaptiveDecrypter.h]:
```suggestion
  virtual const char* GetSessionId() { return nullptr; }
```

File: wvdecrypter/wvdecrypter.cpp [hunk 13]
```diff
@@ -172,7 +172,11 @@ class WV_CencSingleSampleDecrypter : public Adaptive_CencSingleSampleDecrypter
   bool HasKeyId(const uint8_t *keyid);
 
   virtual AP4_Result SetFragmentInfo(AP4_UI32 pool_id, const AP4_UI08 *key, const AP4_UI08 nal_length_size,
-    AP4_DataBuffer &annexb_sps_pps, AP4_UI32 flags)override;
+                                     AP4_DataBuffer& annexb_sps_pps,
+                                     AP4_UI32 flags,
+                                     AP4_UI08 cryptoType,
+                                     AP4_UI08 cryptBlocks,
+                                     AP4_UI08 skipBlocks) override;
```
Comment (CastagnaIT) [inline wvdecrypter/wvdecrypter.cpp]: please clang format it

**Review by CastagnaIT — COMMENTED** [review]
i made a fast test with n€tflix vp9.2 and works
i will try av1 later i have no more time atm

File: src/common/AdaptiveDecrypter.h (hunk 9 above)
Comment (glennguy) [reply src/common/AdaptiveDecrypter.h]: I've gone one better and now send the whole `CryptoInfo` struct. This should have been simple but there were circular includes which took some time to sort out.

**Comment by glennguy** [comment]
@CastagnaIT I think it's ready to look at again when you're ready

File: src/CryptoInfo.h [hunk 14]
```diff
@@ -0,0 +1,17 @@
+/*
```
Comment (CastagnaIT) [inline src/CryptoInfo.h]: this new file does not much sense
because include the CryptoMode
and both files have a single struct
its better merge CryptoInfo.h with CryptoMode.h in a single file something like CryptoUtils

File: src/common/AdaptiveCencSampleDecrypter.cpp [hunk 15]
```diff
@@ -0,0 +1,59 @@
+/*
+ *  Copyright (C) 2023 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#pragma once
```
Comment (CastagnaIT) [inline src/common/AdaptiveCencSampleDecrypter.cpp]: this and the other one added to wvdecrypter*.cpp
looks like a workarounds maybe due to circular includes

File: src/common/AdaptiveCencSampleDecrypter.cpp [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+
+#include "AdaptiveCencSampleDecrypter.h"
+#include "AdaptiveDecrypter.h"
+
+#include "../utils/log.h"
```
Comment (CastagnaIT) [inline src/common/AdaptiveCencSampleDecrypter.cpp:14]: to be removed

File: src/common/AdaptiveCencSampleDecrypter.h [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+ *  Copyright (C) 2023 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#pragma once
+
+#include "../utils/log.h"
```
Comment (CastagnaIT) [inline src/common/AdaptiveCencSampleDecrypter.h]: to be removed

File: src/common/AdaptiveDecrypter.h [hunk 18]
```diff
@@ -8,25 +8,18 @@
...
 #pragma once
 
-#include "../CryptoMode.h"
-
 #include <stdexcept>
 #include <string_view>
 
 #include <bento4/Ap4.h>
 
+struct CryptoInfo;
```
Comment (CastagnaIT) [inline src/common/AdaptiveDecrypter.h]: please add a comment above
`// Forward`

Comment (CastagnaIT) [reply src/common/AdaptiveDecrypter.h]: take in account that if you add a forward other files that will include this header must add also the header that include CryptoInfo struct, since the header that contains CryptoInfo is not big add this forward probably does not lighten the compiler

File: wvdecrypter/wvdecrypter.cpp [hunk 19]
```diff
@@ -529,6 +535,8 @@ void WV_CencSingleSampleDecrypter::GetCapabilities(const uint8_t* key, uint32_t
   {
     AP4_UI32 poolid(AddPool());
     fragment_pool_[poolid].key_ = key ? key : reinterpret_cast<const uint8_t*>(keys_.front().keyid.data());
+    fragment_pool_[poolid].m_cryptoInfo = CryptoInfo();
```
Comment (CastagnaIT) [inline wvdecrypter/wvdecrypter.cpp]: AddPool construct FINFO
so m_cryptoInfo is already constructed so add `= CryptoInfo()`
its not needed

File: src/common/AdaptiveCencSampleDecrypter.cpp [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+      if (AP4_FAILED(result))
+        return result;
+    }
+
+    // decrypt the sample
+    Adaptive_CencSingleSampleDecrypter* decrypter =
+        dynamic_cast<Adaptive_CencSingleSampleDecrypter*>(m_SingleSampleDecrypter);
+    if (!decrypter)
+    {
+      LOG::Log(LOGERROR, "Failed to cast AP4 decrypter to Adaptive");
```
Comment (CastagnaIT) [inline src/common/AdaptiveCencSampleDecrypter.cpp]:
```suggestion
      LOG::LogF(LOGERROR, "Failed to cast AP4 decrypter to Adaptive");
```

**Review by CastagnaIT — APPROVED** [review]
now looks very good
remember to do bento4 release and fix relative commit before merge the PR
